### PR TITLE
Freeze player until chunk is loaded

### LINKED ISSUE #2316: Implement vanilla style falling prevention when the current chunk is not loaded
Cuberite should prevent entities from falling through blocks in unloaded chunks.
This is because entities, or at least players slowly fall when the chunks they are in are not loaded.
Vanilla does this. It is possible to save the location of an entity when entering an unloaded chunk, and teleporting him back when the chunk gets loaded.

This will fix a lot of nether portal problems we currently have, but it's not just portals, it's always when an entity enters an unloaded chunk.

http://forum.mc-server.org/showthread.php?tid=2044&pid=21678#pid21678

### PR DESCRIPTION
**Pull request by SafwatHalaby**
This should prevent players sinking into ground on portal teleportation on slow devices like the RaspberryPi. I did not test this. Can someone try it out? @jammet ?

closes #2316

### CODE REVIEW

File: src/Entities/Player.cpp [hunk 1]
```diff
@@ -220,8 +222,21 @@ void cPlayer::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 
 	m_Stats.AddValue(statMinutesPlayed, 1);
 
+	if (m_IsPositionLocked)
+	{
+		if (a_Chunk.IsValid())
+		{
+			m_IsPositionLocked = false;
+		}
+		else
```
Comment (SafwatHalaby) [inline src/Entities/Player.cpp]: Should I broadcast here?

**Comment by jammet** [comment]
Yeah, I will live to test it :). It's family time now, but I am going to do
that tomorrow night. All the very best for you, all of you! Much love and
happy holidays!

**Comment by SafwatHalaby** [comment]
Merry Christmas :)

**Comment by NiLSPACE** [comment]
Perhaps add two methods `LockPosition(Vector3d)` and `UnlockPosition()` so that plugins can use it? :)

**Comment by SafwatHalaby** [comment]
Yes. But let me get it into a workable condition first :)
The players are not able to move at all (And I'm seeing pitch black all around). Any idea why?

**Comment by NiLSPACE** [comment]
Isn't that the whole point though? That they can't move until the chunk is loaded? That you're seeing black all around is probably the void. If you were a little bit higher (for example Y=100) it would probably look better.

**Comment by SafwatHalaby** [comment]
They can't move under normal conditions. I don't think my chunks are unloaded, I haven't changed anything or unloaded on purpose.

File: src/Entities/Player.cpp [hunk 2]
```diff
@@ -63,6 +63,7 @@ cPlayer::cPlayer(cClientHandlePtr a_Client, const AString & a_PlayerName) :
 	m_GameMode(eGameMode_NotSet),
 	m_IP(""),
 	m_ClientHandle(a_Client),
+	m_IsPositionLocked(true),
```
Comment (NiLSPACE) [inline src/Entities/Player.cpp]: You're locking players by default?

**Comment by SafwatHalaby** [comment]
It's working fine with the latest fix :)
Is there a way to unload a chunk?

File: src/Entities/Player.cpp (hunk 2 above)
Comment (SafwatHalaby) [inline src/Entities/Player.cpp]: Yes. Locked until proven otherwise ;)

**Comment by NiLSPACE** [comment]
I'm affraid not. What you can do is teleport to not-generated chunks using `/tp ~ ~ ~750`

**Comment by SafwatHalaby** [comment]
I need a slower device. Guess we'll wait for Jammet to test it on a Raspberry Pi.

File: src/Entities/Player.h [hunk 3]
```diff
@@ -138,6 +138,15 @@ class cPlayer :
 
 	// tolua_begin
 
+	/** Prevent the palyer from moving and lock him into a_Location. */
+	void Freeze(const Vector3d & a_Location);
```
Comment (SafwatHalaby) [inline src/Entities/Player.h]: Will this be enough for plugin export? I just added  it inside the `tolua_begin` are. I never exported commands before.

Comment (NiLSPACE) [inline src/Entities/Player.h]: Yes, that should do it. Don't forget to document the functions though :)

**Comment by SafwatHalaby** [comment]
Rebased.

File: src/Entities/Player.h (hunk 3 above)
Comment (madmaxoft) [inline src/Entities/Player.h]: This should be enough. However, I question the parameter, why is a position needed? Shouldn't the current position be the one?

**Comment by madmaxoft** [comment]
You could always test it by putting a Sleep in the ChunkSender. Send two chunks per second and you're bound to see some results :)

**Comment by SafwatHalaby** [comment]
@madmaxoft Will do.

Is my broadcasting  code correct?

File: src/Entities/Player.cpp [hunk 4]
```diff
@@ -220,8 +222,30 @@ void cPlayer::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
...
+		}
+		else
+		{
+			SetPosition(m_FrozenPosition);
+
+			SetSpeed(0, 0, 0);  // Todo send every second or so, not every tick.
+			if (m_FreezeCounter % 20 == 0)
+			{
+				BroadcastMovementUpdate(m_ClientHandle.get());
+				m_ClientHandle->SendPlayerPosition();
```
Comment (SafwatHalaby) [inline src/Entities/Player.cpp]: Is this correct?

Comment (madmaxoft) [inline src/Entities/Player.cpp]: Seems so, I guess it's alright.

**Comment by SafwatHalaby** [comment]
Is freezing until the chunk is loaded enough? What about waiting until the client receives the chunk?

**Comment by madmaxoft** [comment]
We can never know when the chunk is really received, so there's no way to implement that anyway. This should be enough to cover absolute most of the cases.

**Comment by SafwatHalaby** [comment]
@madmaxoft What about freezing for 1 extra second after the chunk is loaded?

**Comment by madmaxoft** [comment]
That would be bothersome to players and servers on good hardware. Let's just keep it this way.

**Comment by SafwatHalaby** [comment]
lgtm. Done *some* testing.

**Comment by madmaxoft** [comment]
I never know if this "lgtm" is a "let's get this merged" or "looks good to me" :P

**Comment by SafwatHalaby** [comment]
Looks good to me! :P

File: Server/Plugins/APIDump/APIDesc.lua [hunk 5]
```diff
@@ -1870,6 +1870,7 @@ a_Player:OpenWindow(Window);
 				Feed = { Params = "AddFood, AddSaturation", Return = "bool", Notes = "Tries to add the specified amounts to food level and food saturation level (only positive amounts expected). Returns true if player was hungry and the food was consumed, false if too satiated." },
 				FoodPoison = { Params = "NumTicks", Return = "", Notes = "Starts the food poisoning for the specified amount of ticks; if already foodpoisoned, sets FoodPoisonedTicksRemaining to the larger of the two" },
 				ForceSetSpeed = { Params = "{{Vector3d|Direction}}", Notes = "Forces the player to move to the given direction." },
+				Freeze = { Params = "{{Vector3d|Location}}", Return = "", Notes = "Prevent the player from moving and lock him into Location until unfreeze() is called" },
```
Comment (madmaxoft) [inline Server/Plugins/APIDump/APIDesc.lua]: `Prevents`; describe what happens if a wild-away position is used - is the player teleported?

Comment (madmaxoft) [inline Server/Plugins/APIDump/APIDesc.lua]: Is the freezing stacking? If two plugins call Freeze and one calls Unfreeze, will the player still be frozen?

File: src/Entities/Player.cpp [hunk 6]
```diff
@@ -220,8 +222,29 @@ void cPlayer::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
...
+			SetSpeed(0, 0, 0);
+			if (m_FreezeCounter % 60 == 0)
+			{
+				BroadcastMovementUpdate(m_ClientHandle.get());
+				m_ClientHandle->SendPlayerPosition();
+			}
+			return;
+		}
+	}
+
```
Comment (madmaxoft) [inline src/Entities/Player.cpp]: This new block of code could use some comment - why it's there, what's it doing (high-level)

File: src/Entities/Player.h [hunk 7]
```diff
@@ -576,6 +588,18 @@ class cPlayer :
 
 	cSlotNums m_InventoryPaintSlots;
 
+	/** If a chunk is not loaded, we lock m_Location to this position. */
```
Comment (madmaxoft) [inline src/Entities/Player.h]: -> `If m_IsFrozen is true`

File: src/Entities/Player.h [hunk 8]
```diff
@@ -576,6 +588,18 @@ class cPlayer :
...
+	Vector3d m_FrozenPosition;
+
+	/** If true, we are locking m_Position to m_FrozenPosition. */
+	bool m_IsFrozen;
+
+	/** */
+	int m_FreezeCounter;
+
+	/** Was the player forzen manually by a plugin or automatically by the server? */
```
Comment (madmaxoft) [inline src/Entities/Player.h]: `frozen`

File: src/Entities/Player.h [hunk 9]
```diff
@@ -661,6 +685,9 @@ class cPlayer :
 	/** Tosses a list of items. */
 	void TossItems(const cItems & a_Items);
 
+	/** Pins the player to a_Location until Unfreeze() is called. if ManuallyFrozen is false, the player will unfreeze when the chunk is loaded. */
```
Comment (madmaxoft) [inline src/Entities/Player.h]: Break line before `If`

File: Server/Plugins/APIDump/APIDesc.lua (hunk 5 above)
Comment (SafwatHalaby) [inline Server/Plugins/APIDump/APIDesc.lua]: Assuming SetPosition and `ClientHandle->SendPlayerPosition` work for any arbitary distance, then yes, he is teleported.

**Comment by madmaxoft** [comment]
Well of course it looks good to you, you're the one who wrote it :P
Fix my comments and it'll be good from me, too. Wait for a third set of eyes, @worktycho usually spots those hard-to-think-about subtle bugs.

**Comment by SafwatHalaby** [comment]
Changes applied + squashed.

File: src/Entities/Player.cpp (hunk 6 above)
Comment (SafwatHalaby) [inline src/Entities/Player.cpp]: Done.

**Comment by SafwatHalaby** [comment]
OSX1 build failed. Reason:
```
CMake Warning (dev) at /usr/local/Cellar/cmake/3.4.0/share/cmake/Modules/CheckSymbolExists.cmake:58 (if):Slave went offline during the build
ERROR: Connection was broken: java.io.IOException: Unexpected termination of the channel
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:50)
Caused by: java.io.EOFException
	at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2332)
	at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:2801)
	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:801)
	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:299)
	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:48)
	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:48)

Build step 'Execute shell' marked build as failure
ERROR: Step ‘Archive the artifacts’ failed: no workspace for Cuberite OSX x64 Pull Requests #57
Setting status of 8e3d055ac569a8c0b941587c4f87d65ada24e914 to FAILURE with url https://builds.cuberite.org/job/Cuberite%20OSX%20x64%20Pull%20Requests/57/ and message: 'Build Failed
 No test results found.'
Using context: cuberite-buildserver/pr
Finished: FAILURE
``

**Comment by SphinxC0re** [comment]
I restarted the node. Soon the OS X node will have cable-network access. ATM it only has WiFi

**Comment by SafwatHalaby** [comment]
I think the OSX check should be removed from Github until its issues are resolved.

**Comment by SphinxC0re** [comment]
Would you please rebase again

**Comment by SafwatHalaby** [comment]
Done.

**Comment by jammet** [comment]
I'll be back home soon, and then you were test this on the Pi! Much love
and thanks so much for doing this! I really appreciate that! :)

**Comment by jammet** [comment]
I've been home for half an hour and got right to it. The build I am working
with, is:

Cuberite Jenkins gcc armhf Release (master) build id: #320 from commit id:
6a22605fae80becb12318785389d07974a35cbbb built at: Fri 25 Dec 14:05:00 CET
2015

I've an in and out of gates, and teleported around. What I can say is -
although this could be the server being in a much better mood since many
months - gate travel works *better* than in previous builds. I've been able
to run around, through various different nether gates. Back and forth. Back
and forth. Pick another, back and forth, pick another, and so on.

I did get the gate loopback a few times. If I see the world as it is being
loaded, sometimes it looks like it's loading the normal world, but with
nether textures, or it looks like the nether with structures built in the
normal world. But I had that effect before, so that just plain didn't
change. What I also still notice that the character does not seem to be
frozen. He's still slooowing sinking. And if the server loads the world too
slowly, I end up in the ground and suffocate as before.

Flying around at really fast speeds, so I find a lot of chunks that load
slow, I found that I could still just fly through unloaded chunks. Can
still move around in there, as before. Meaning, so far, it never looks like
the player is actually frozen, or I just can't tell.

However, as said before, traveling through gates works more reliable now
than it ever has, but the problems of suffocation and looping are still
here. I'll keep trying new builds over the next days and post what I find.

Also -- I connect locally. Please feel free to see how this works for you
as a remote player. The server is still at jammet.duckdns.org. :)

Thanks so much for all the hard work! I look forward to this finally being
checked off the list. For lower power servers, this bit is very important!

**Comment by SafwatHalaby** [comment]
Thanks for the feedback!
Note that this PR is only designed to deal with one of the described problems: Freezing the a teleported player till the chunk is loaded. It doesn't deal with flying players or the loopback issue.
I'll check out why players slowly sink.

**Comment by jammet** [comment]
Also -- I connect locally. Please feel free to see how this works for you
as a remote player. The server is still at jammet.duckdns.org. :)

I'll keep testing :).
Thanks so much for all the hard work! I look forward to this finally being
checked off the list. For lower power servers, this bit is very important!

**Comment by SafwatHalaby** [comment]
Mind giving me a portal coords @ your server?

**Comment by SafwatHalaby** [comment]
I submitted a patch which should fix the sinking issue.

**Comment by SafwatHalaby** [comment]
Actually, the fix is useless.

**Comment by jammet** [comment]
If you go to the spawn, there's a portal there :). Even if you don't need the info anymore.
The coordinates are: ... oh you're on the server,...


Now this happened:

[16:32:55] Player Safwat has joined the game
[16:33:23] Player Safwat issued an unknown command: "/worls"
[16:33:26] Player Safwat tried to execute forbidden command: "/worlds"
say Coming
[16:48:58] Executing console command: "say Coming"
[16:49:05] [Safwat]: Hello.
[16:49:16]   D:    | Cuberite has encountered an error and needs to close
[16:49:16] Details | SIGSEGV: Segmentation fault
[16:49:16] Cuberite Jenkins gcc armhf Release (master) build id: #320
[16:49:16] from commit id: 6a22605fae80becb12318785389d07974a35cbbb built at: Fri 25 Dec 14:05:00 CET 2015
./Cuberite(_Z15PrintStackTracev+0x14)[0x1fc390]
./Cuberite(_Z14NonCtrlHandleri+0xd0)[0x1e9460]
/lib/arm-linux-gnueabihf/libc.so.6(__default_sa_restorer_v2+0x0)[0x76c6cb20]
./Cuberite(_ZN24cBlockingSslClientSocket10DisconnectEv+0x20)[0x42c714]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
./Cuberite(_ZN10cMojangAPI13SecureRequestERKSsS1_RSs+0x80)[0x2f5f2c]
[16:49:16]   D:    | Cuberite has encountered an error and needs to close
[16:49:16] Details | SIGABRT: Server self-terminated due to an internal fault
[16:49:16] Cuberite Jenkins gcc armhf Release (master) build id: #320
[16:49:16] from commit id: 6a22605fae80becb12318785389d07974a35cbbb built at: Fri 25 Dec 14:05:00 CET 2015
./Cuberite(_Z15PrintStackTracev+0x14)[0x1fc390]
./Cuberite(_Z14NonCtrlHandleri+0xd0)[0x1e9460]
/lib/arm-linux-gnueabihf/libc.so.6(__default_sa_restorer_v2+0x0)[0x76c6cb20]
/lib/arm-linux-gnueabihf/libc.so.6(gsignal+0x40)[0x76c6b8dc]
./start.sh: line 1: 15934 Aborted                 ./Cuberite

**Comment by tigerw** [comment]
I see the rationale behind this, but the player is not meant to exit the loading screen on cross-world travel before spawn chunks in the destination have been loaded.

Perhaps investigate why that doesn't happen rather than prevent movement in unloaded chunks. When server-sided collision checks are implemented, this will become unnecessary and another deprecated but exported (so non removable, ugh) function.

**Comment by tigerw** [comment]
Um, so with regards to the above:

A respawn packet when sent causes the client to exit the holding screen and start rendering the world. I initially sent this essentially instantaneously when world traversal was requested, because InitializeWorld blocked and would cause the client to disconnect prematurely. Therefore in most cases the chunk loading will not catch up before Steve falls through.

I propose the changes to be adapted to a ChunkStay for the destination chunk (or 3×3 chunks if appropriate), which upon completion would send a respawn packet.

**Comment by SafwatHalaby** [comment]
> Perhaps investigate why that doesn't happen rather than prevent movement in unloaded chunks.
I think this prevention is needed anyways for e.g. preventing flying too deep into ungenerated areas and chocking a raspberry pi.

I'll need to look at the code a bit and understand what you mean by ChunkStay to make a comment on the rest of the things you mentioned.

By the way, testing on Jammet's server made me  \* suspect \* the chunk is never returning "false" for `isValid()` and therefore the freeze code is never executed. (Perhaps loaded chunks never move their entities to unloaded neighbors?)

**Comment by SafwatHalaby** [comment]
@jammet just making sure: When we performed the tests, was your server on the master branch or this branch? Because the freeze code was never executed.

**Comment by jammet** [comment]
The master branch ; I hope that was correct to do? I always update to the
master branch, because I thought that is where all the patches go first.

**Comment by SafwatHalaby** [comment]
That explains why my code was never executed. It was never there ;)

**Comment by SafwatHalaby** [comment]
**Now** the fix is in the master branch.

**Comment by jammet** [comment]
I'm so sorry I messed this up for you.

**Comment by jammet** [comment]
I've downloaded the most recent build, and if it's really in there now, it
does not seem to be doing what you want it to. I'll do my very best to help
you test it. If, for some reason, I downloaded a build where this wasn't
included, yet, then I apologise and will try again tomorrow.

[21:11:54] Cuberite Jenkins gcc armhf Release (master) build id: #325
[21:11:54] from commit id: bfafc39576f4b6a2c0d5017795ba420fdd3625c4 built
at: Sat 26 Dec 20:12:22 CET 2015

That's the build I'm using. In empty chunks, the player still slowly sinks
into the ground.

**Comment by tigerw** [comment]
...but, my objections?

**Comment by SafwatHalaby** [comment]
I'm sorry @tigerw , I just wanted to test this on Jammet's server asap and Jammet only knows how to compile Master. I did not ignore your comments, and will make more PRs if needed.

**Comment by tigerw** [comment]
It doesn't matter, just hoping that xoft doesn't mind if the exported functions get changed later :P

**Comment by jammet** [comment]
I'm not compiling these, I run a script:

mv Cuberite.tar.gz previousCuberite.tar.gz
wget
http://builds.cuberite.org/job/Cuberite%20Linux%20raspi-armhf%20Master/lastSuccessfulBuild/artifact/Cuberite.tar.gz

And just unpack. ^^'.

**Comment by LogicParrot** [comment]
Self-assigning in order to not forget checking out @tigerw 's proposal.

**Comment by LogicParrot** [comment]
>Perhaps investigate why that doesn't happen rather than prevent movement in unloaded chunks. When server-sided collision checks are implemented, this will become unnecessary and another deprecated but exported (so non removable, ugh) function.

`freeze()` is useful regardless of this though.

**Comment by LogicParrot** [comment]
>this will become unnecessary and another deprecated but exported (so non removable, ugh) function.

`freeze()` is useful regardless of this though.

Edit: Quoted the right part.

**Pull request merged**